StackExchange.Redis/StackExchange/Redis/ConnectionMultiplexer.cs

```diff
@@ -2058,37 +2069,60 @@ internal T ExecuteSyncImpl<T>(Message message, ResultProcessor<T> processor, Ser
             }
         }
 
+        private static void AppendTimeoutHelpLinkUrl(StringBuilder sb)
+        {
+            sb.Append(" Please take a look at this article for some tips for investigating timeouts: ");
+            sb.Append(timeoutHelpLink);
+            sb.Append(".");
+        }
+
 #if !CORE_CLR
-        private static string GetSystemCpuPercent()
+        /// <summary>
+        /// Gathers system stats and appends detailed info about possible causes of timeouts to the error message StringBuilder
+        /// </summary>
+        private static void AddTimeoutRootCauseIfPossible(StringBuilder sb, ServerEndPoint server, out ThreadPoolStats iocp, out ThreadPoolStats worker, out string localCpuPercent)
         {
+            localCpuPercent = "unavalable";
+            ThreadPoolStats.Gather(out iocp, out worker);
+
             float systemCPU;
-            if (PerfCounterHelper.TryGetSystemCPU(out systemCPU))
+            bool gotCpuUsageSuccessfully = PerfCounterHelper.TryGetSystemCPU(out systemCPU);
+
+            if (server.ConnectionState != PhysicalBridge.State.ConnectedEstablished)
             {
-                return Math.Round(systemCPU, 2) + "%";
+                sb.Append($" The connection to the server was lost and was not restored in the timeout specified.");
+                return; // don't add IOCP and CPU details to the error message because connection management trumps all other causes of timeouts.
+            }
+
+            bool detailsAdded = false;
+            bool timeoutHelpLinkAdded = false;
+            if (iocp.Busy > iocp.Min || worker.Busy > worker.Min)

```

Comment:
The goal of the check in is to give customers who don't have any experience with debugging these types of errors a better clue as to where to start.  We see many customers come to us to help them diagnose timeouts.  Customers typically fall into two buckets - those who understand how the ThreadPool scales and those that don't.  

For those that do understand this, then a false lead is something they probably have sufficient skill to diagnose themselves and they will be able to look past any false lead.  

However, for those that don't understand the thread pool very well, this is a great hint at a common cause of timeouts.  We help many customers per week adjust their threadpool settings and this fixes the issue for almost all of them.

I would argue that even for a high concurrency system, you should probably adjust your threadpool settings to avoid the growth throttling costs so that you never hit these types of timeouts...